src/merge_operator.rs

```diff
@@ -0,0 +1,373 @@
+            &mut iterator,
+            vec![
+                RowEntry::new_merge(b"key1", b"4321", 4),
+                RowEntry::new_value(b"key2", b"1", 5),
+                RowEntry::new_value(b"key3", b"321", 8),
+            ],
+        )
+        .await;
+    }
+
+    #[derive(Debug)]
+    struct TestCase {
+        unsorted_data: Vec<RowEntry>,
+        expected: Vec<RowEntry>,
+        merge_different_expire_ts: bool,
+    }
+
+    impl Default for TestCase {
+        fn default() -> Self {
+            Self {
+                unsorted_data: vec![],
+                expected: vec![],
+                merge_different_expire_ts: true,
+            }
+        }
+    }
+
+    #[rstest]
+    #[case::different_expire_ts_read_path(TestCase {
+        unsorted_data: vec![
+            RowEntry::new_merge(b"key1", b"1", 1).with_expire_ts(1),
+            RowEntry::new_merge(b"key1", b"2", 2).with_expire_ts(2),
+            RowEntry::new_merge(b"key1", b"3", 3).with_expire_ts(3),
+            RowEntry::new_value(b"key2", b"1", 4),
+            RowEntry::new_merge(b"key3", b"1", 5).with_expire_ts(1),
+            RowEntry::new_merge(b"key3", b"2", 6).with_expire_ts(1),
+            RowEntry::new_merge(b"key3", b"3", 7).with_expire_ts(2),
+        ],
+        expected: vec![
+            RowEntry::new_merge(b"key1", b"321", 3).with_expire_ts(1),
+            RowEntry::new_value(b"key2", b"1", 4),
+            RowEntry::new_merge(b"key3", b"321", 7).with_expire_ts(1),
+        ],
+        ..TestCase::default()
+    })]
+    #[case::different_expire_ts_write_path(TestCase {
+        unsorted_data: vec![
+            RowEntry::new_merge(b"key1", b"1", 1).with_expire_ts(1),
+            RowEntry::new_merge(b"key1", b"2", 2).with_expire_ts(2),
+            RowEntry::new_merge(b"key1", b"3", 3).with_expire_ts(3),
+            RowEntry::new_value(b"key2", b"1", 4),
+            RowEntry::new_merge(b"key3", b"1", 5).with_expire_ts(1),
+            RowEntry::new_merge(b"key3", b"2", 6).with_expire_ts(1),
+            RowEntry::new_merge(b"key3", b"3", 7).with_expire_ts(2),
+        ],
+        expected: vec![
+            RowEntry::new_merge(b"key1", b"3", 3).with_expire_ts(3),
+            RowEntry::new_merge(b"key1", b"2", 2).with_expire_ts(2),
+            RowEntry::new_merge(b"key1", b"1", 1).with_expire_ts(1),
+            RowEntry::new_value(b"key2", b"1", 4),
+            RowEntry::new_merge(b"key3", b"3", 7).with_expire_ts(2),
+            RowEntry::new_merge(b"key3", b"21", 6).with_expire_ts(1),
+        ],
+        // On write path (compaction, memtable), we don't merge entries
+        // with different expire timestamps to allow per-element expiration.
+        merge_different_expire_ts: false
+    })]
+    #[case::merge_with_tombstone(TestCase {
+        unsorted_data: vec![
+            RowEntry::new_merge(b"key1", b"1", 1),
+            RowEntry::new_merge(b"key1", b"2", 2),
+            RowEntry::new_tombstone(b"key1", 3),
```

Comment:
now that we're returning each version of a record from the iterators, shouldn't we be returning tombstones as well? otherwise how is a user of the iterator supposed to observe that a value was deleted? e.g. if I just had:
```
unsorted_data: vec![
            RowEntry::new_merge(b"key1", b"1", 1),
            RowEntry::new_merge(b"key1", b"2", 2),
            RowEntry::new_tombstone(b"key1", 3),
]
```
Then, the user would get back key1->21 when iterating but would have no way to observe that this value is in fact deleted.

Reply:
This is correct, because we're turning `tombstone + merge` into a `value` (as per-rfc). `value` already contains information that all previous "entries" should be invalidated.

Reply:
https://github.com/slatedb/slatedb/blob/main/rfcs/0006-merge-operator.md#write-path--compactions

Reply:
Yep, this makes sense to me now. Thanks!